addons/mail/models/mail_scheduled_message.py

```diff
@@ -0,0 +1,262 @@
+    _name = 'mail.scheduled.message'
+    _description = 'Scheduled Message'
+
+    # content
+    subject = fields.Char('Subject')
+    body = fields.Html('Contents', sanitize_style=True)
+    scheduled_date = fields.Datetime('Scheduled Date', required=True)
+    attachment_ids = fields.Many2many(
+        'ir.attachment', 'scheduled_message_attachment_rel',
+        'scheduled_message_id', 'attachment_id',
+        string='Attachments')
+
+    # related document
+    model = fields.Char('Related Document Model', required=True)
+    res_id = fields.Many2oneReference('Related Document Id', model_field='model', required=True)
+
+    # origin
+    author_id = fields.Many2one('res.partner', 'Author', required=True)
+
+    # recipients
+    partner_ids = fields.Many2many('res.partner', string='Recipients')
+
+    # characteristics
+    is_note = fields.Boolean('Is a note', default=False, help="If the message will be posted as a Note.")
+
+    # notify parameters (email_from, mail_server_id, force_email_lang,...)
+    notification_parameters = fields.Text('Notification parameters')
+
+    @api.constrains('scheduled_date')
+    def _check_scheduled_date(self):
+        if any(scheduled_message.scheduled_date < fields.Datetime().now() for scheduled_message in self):
+            raise ValidationError(_("A Scheduled Message cannot be scheduled in the past"))
+
+    # ------------------------------------------------------
+    # CRUD / ORM
+    # ------------------------------------------------------
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        # check that messages can be posted on the models
+        for model in {vals['model'] for vals in vals_list}:
+            if not (model in self.pool and issubclass(self.pool[model], self.pool['mail.thread'])):
+                raise UserError(_("A message cannot be scheduled on a model that does not have a mail thread."))
+
+        Scheduled = self.browse()
+        for vals in vals_list:
+            # make sure user can post on the related records
+            Scheduled._check(vals)
+
+        scheduled_messages = super().create(vals_list)
+        # transfer attachments from composer to scheduled messages
+        for scheduled_message in scheduled_messages:
+            if attachments := scheduled_message.attachment_ids:
+                attachments.filtered(
+                    lambda a: a.res_model == 'mail.compose.message' and not a.res_id and a.create_uid.id == self.env.uid
+                ).write({
+                    'res_model': scheduled_message._name,
+                    'res_id': scheduled_message.id,
+                })
+        # schedule cron trigger
+        if scheduled_messages:
+            self.env.ref('mail.ir_cron_post_scheduled_message')._trigger_list(
+                set(scheduled_messages.mapped('scheduled_date'))
+            )
+        return scheduled_messages
+
+    @api.model
+    def _search(self, domain, offset=0, limit=None, order=None):
```

Comment:
Don't you also have to override ``fetch`` and ``_read_group`` to call our custom check ? Otherwise those call standard access check on read, which is globally allow all.

Reply:
Seems those make _search queries.